jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponse.java

```diff
@@ -249,136 +271,162 @@ protected void onCompleteFailure(Throwable x)
-                            slice = _copy;
-                            BufferUtil.append(_copy, content);
-                        }
-
-                        // transfer the data from the slice to the deflater
-                        byte[] array = slice.array();
-                        int off = slice.arrayOffset() + slice.position();
-                        int len = slice.remaining();
-                        _crc.update(array, off, len);
-                        // Ideally we would want to use the ByteBuffer API for Deflaters. However due the the ByteBuffer implementation
-                        // of the CRC32.update() it is less efficient for us to use this rather than to convert to array ourselves.
-                        _deflaterEntry.get().setInput(array, off, len);
-                        slice.position(slice.position() + len);
-                        if (_last && _content == null && BufferUtil.isEmpty(content))
-                            deflater.finish();
+                        BufferUtil.flipToFlush(outputBuffer, 0);
+                        write(false, outputBuffer);
+                        return Action.SCHEDULED;
                     }
                 }
+            }
 
-                // deflate the content into the available space in the buffer
-                int off = _buffer.arrayOffset() + _buffer.limit();
-                int len = BufferUtil.space(_buffer);
-                int produced = deflater.deflate(_buffer.array(), off, len, _syncFlush ? Deflater.SYNC_FLUSH : Deflater.NO_FLUSH);
-                _buffer.limit(_buffer.limit() + produced);
+            if (_last)
+            {
+                _state.set(GZState.FINISHING);
+                deflater.finish();
+                return finishing(deflater, outputBuffer);
             }
 
-            // If we have finished deflation and there is room for the trailer.
-            if (deflater.finished() && BufferUtil.space(_buffer) >= 8)
+            BufferUtil.flipToFlush(outputBuffer, 0);
+            if (outputBuffer.hasRemaining())
             {
-                // add the trailer and recycle the deflator to flag that we will have had completeSuccess when
-                // the write below completes.
-                addTrailer();
-                _deflaterEntry.release();
-                _deflaterEntry = null;
+                write(false, outputBuffer);
+                return Action.SCHEDULED;
+            }
+
+            // the content held by GzipBufferCB is fully consumed as input to the Deflater instance, we are done
+            if (BufferUtil.isEmpty(_content))
+                return Action.SUCCEEDED;
+
+            // No progress made on deflate, but the _content wasn't consumed, we shouldn't be able to reach this.
+            throw new AssertionError("No progress on deflate made for " + this);
+        }
+
+        private Action finishing(Deflater deflater, ByteBuffer outputBuffer)
```

Comment:
Add javadoc to say something like: "this method is called by `compressing`, once the last chunk is compressed; or directly from `process` if an earlier call to `finishing` was unable to complete.